### PR TITLE
Refine use final sort.

### PR DESCRIPTION
**Pull request by claygoddard**
For best results, refine should really sort by the final estimated sort before deciding which buckets to use for refinement.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java [hunk 1]
```diff
@@ -232,7 +232,9 @@ public Map<String, Object> getRefinement(Context mcontext) {
       return null;
     }
 
-    final FacetRequest.FacetSort initial_sort = null == freq.prelim_sort ? freq.sort : freq.prelim_sort;
+    // Use final sort here (as opposed to prelim_sort if available) to determine top buckets for refinement.
+    // At this point, we want to take our best guess at the final sorted values for refinement.
+    final FacetRequest.FacetSort sort = freq.sort;
```
Comment (claygoddard) [inline solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:237]: Maybe this should be an option if/when upstreamed? But I truly feel that this is _far_ better here. Refinement is already operating on results that were prelim sorted, truncated, then resorted on a per shard basis. It should pick which buckets to refine based on estimated final sorting when we are working at the merge level. Thoughts?

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java [hunk 2]
```diff
@@ -259,9 +261,9 @@ public Map<String, Object> getRefinement(Context mcontext) {
 
         // when we don't have to worry about mincount pruning, there is no need for any
         // over refinement for these sorts..
-        if (freq.mincount <= 1 && ("index".equals(initial_sort.sortVariable)
-                                   || ("count".equals(initial_sort.sortVariable)
-                                       && FacetRequest.SortDirection.desc == initial_sort.sortDirection))) {
+        if (freq.mincount <= 1 && ("index".equals(sort.sortVariable)
+                                   || ("count".equals(sort.sortVariable)
+                                       && FacetRequest.SortDirection.desc == sort.sortDirection))) {
           // No-Op
```
Comment (claygoddard) [inline solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:267]: I really don't understand this no-op. It's just under the heuristic approach but it does seem like index / count sorting could certainly still benefit from a little overrefinement.. but leaving it for now.

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetRefinement.java [hunk 3]
```diff
@@ -555,14 +556,14 @@ public void testPrelimSortingWithRefinement() throws Exception {
...
+                  + "            {val:X,count:4},"
                   + "  ] },"
                   + "  cat_2:{ buckets:[ "
-                  + "            {val:X,count:4,sum_p:4.0}," // count desc
-                  + "            {val:A,count:1,sum_p:1.0}," // index order tie break
-                  + "            {val:B,count:1,sum_p:1.0},"
+                  + "            {val:C,count:6,sum_p:6.0}," // count desc
+                  + "            {val:Y,count:5,sum_p:5.0}," // index order tie break
+                  + "            {val:Z,count:2,sum_p:2.0},"
                   + "  ] }"
```
Comment (claygoddard) [inline solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetRefinement.java:567]: The results in these tests are clearly better relative to request -- but probably would need to reformulate a little for upstream to truly show sorting behavior with regards to refinement.

Comment (magibney) [reply solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetRefinement.java:567]: Agreed this results appear "better"; but as I think you're acknowledging, these tests were specifically designed to surface weird behavior that's possible as a consequence of prelim_sort, so making the results "better" is not the point of these tests. We can still expect counterintuitive behavior in some cases from prelim_sort, it's just no longer triggered by this test as currently designed.

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java (hunk 2 above)
Comment (magibney) [reply solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:267]: For the "index sort" case I think this makes sense (at least for the non-prelim_sort case, and for the prelim_sort case, before the change introduced by this PR)?

For `count desc`, I see what you mean, but I guess for a default the assumption they're making is that there's likely to be enough overlap among top terms that overrefinement is not necessary in addition to overrequest?

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java (hunk 1 above)
Comment (magibney) [reply solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:237]: I _think_ this makes sense?

Trying to play devil's advocate though: I wonder if it depends on the particular kind of sort? Practically speaking, most cases would be prelim_sort by `count desc`. I think `prelim_sort` was initially introduced for sort-by-relatedness (correlation); a little hand-wavy here, but I think `processEmpty` can be important for correlation, filling in background set counts that affect relatedness calculation. It's possible that prematurely calculating `relatedness` based on incomplete information could skew results _more_ than determining refinement based on prelim sort-by-count -- particularly for the case where values might be correlated to particular shards (dependent on shard routing scheme)?

**Review by magibney — COMMENTED** [review]
I'm assuming you saw a substantial improvement here, else you wouldn't have bothered making the change? It makes sense to me, and particularly if you saw improvements for our use case I see no reason not to go ahead.

I raised some concerns wrt making this non-configurable though, mainly relevant to upstreaming.

**Comment by claygoddard** [comment]
> I raised some concerns wrt making this non-configurable though, mainly relevant to upstreaming.

Yeah, in one of my comments I mentioned we should give the option. I'll actually go ahead and add that because it definitely feels use case dependent on what you'd actually want.

**Comment by claygoddard** [comment]
> I'm assuming you saw a substantial improvement here, else you wouldn't have bothered making the change? It makes sense to me, and particularly if you saw improvements for our use case I see no reason not to go ahead.

And yeah, I was seeing refinement result in a worse top N (especially with not overrefining the full overrequest -- which is currently what we are doing, overrefine = 0), even if it is getting accurate results for that top N. But we know in our case prelim sort and sort both are correlated.

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java (hunk 1 above)
Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:237]: Yeah, I think that makes sense as well. I'm just going to throw in an option for this one :).

File: solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java (hunk 2 above)
Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java:267]: I guess. But for items near the limit, there are bound to be differing results from the different shards that did or did not make it into the initial results. But `count desc` is probably better than most for relatively evenly distributed data.

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetRefinement.java (hunk 3 above)
Comment (claygoddard) [reply solr/core/src/test/org/apache/solr/search/facet/TestJsonFacetRefinement.java:567]: Yep yep, agreed these tests were trying to flex very specific behavior even if results are kind of bad. Will look into some updates to show new option behavior (once I also add option flag).

**Comment by claygoddard** [comment]
I have been putting off coming back to this. I think I'm going to close this one and consider implementing for Solr 9. If I need this sooner, I can do Solr 8 too. But we still haven't enabled refinement for anyone.. at this point would probably need some further perf testing as well.

**Pull request closed without merging**